### PR TITLE
Cross field validation #185471858

### PR DESCRIPTION
**Pull request by coltborg**
[Starter app PR](https://github.com/codeforamerica/form-flow-starter-app/pull/198)

### CODE REVIEW

File: src/main/java/formflow/library/ScreenController.java [hunk 1]
```diff
@@ -153,16 +153,16 @@ ModelAndView postScreen(
...
     FormSubmission formSubmission = new FormSubmission(formData);
     actionManager.handleOnPostAction(currentScreen, formSubmission, submission);
 
     // Field validation
-    var errorMessages = validationService.validate(currentScreen, flow, formSubmission);
+    var errorMessages = validationService.validate(currentScreen, flow, formSubmission, submission);
     handleErrors(httpSession, errorMessages, formSubmission);
 
-    if (errorMessages.size() > 0) {
+    if (!errorMessages.isEmpty()) {
```
Comment (coltborg) [inline src/main/java/formflow/library/ScreenController.java:165]: This was a recommendation from IntelliJ.

File: src/main/java/formflow/library/ScreenController.java [hunk 2]
```diff
@@ -279,9 +289,9 @@ ModelAndView updateOrCreateIteration(
           .orElse(null);
     }
 
-    var errorMessages = validationService.validate(currentScreen, flow, formSubmission);
+    var errorMessages = validationService.validate(currentScreen, flow, formSubmission, submission);
     handleErrors(httpSession, errorMessages, formSubmission);
-    if (errorMessages.size() > 0) {
+    if (!errorMessages.isEmpty()) {
```
Comment (coltborg) [inline src/main/java/formflow/library/ScreenController.java:294]: This was a recommendation from IntelliJ.

File: src/main/java/formflow/library/ScreenController.java [hunk 3]
```diff
@@ -474,7 +488,7 @@ private String getNextScreenName(Submission submission,
 
     List<NextScreen> nextScreens = getConditionalNextScreen(currentScreen, submission, subflowUuid);
 
-    if (isConditionalNavigation(currentScreen) && nextScreens.size() > 0) {
+    if (isConditionalNavigation(currentScreen) && !nextScreens.isEmpty()) {
```
Comment (coltborg) [inline src/main/java/formflow/library/ScreenController.java:491]: This was a recommendation from IntelliJ.

File: src/main/java/formflow/library/ScreenController.java [hunk 4]
```diff
@@ -625,7 +639,7 @@ private Map<String, Object> createModel(String flow, String screen, HttpSession
 
   private void handleErrors(HttpSession httpSession, Map<String, List<String>> errorMessages,
       FormSubmission formSubmission) {
-    if (errorMessages.size() > 0) {
+    if (!errorMessages.isEmpty()) {
```
Comment (coltborg) [inline src/main/java/formflow/library/ScreenController.java:642]: This was a recommendation from IntelliJ.

File: src/main/java/formflow/library/config/ActionManager.java [hunk 5]
```diff
@@ -15,7 +15,7 @@
 @Component
 public class ActionManager {
 
-  private HashMap<String, Action> actions = new HashMap<>();
+  private final HashMap<String, Action> actions = new HashMap<>();
```
Comment (coltborg) [inline src/main/java/formflow/library/config/ActionManager.java:18]: This was a recommendation from IntelliJ.

File: README.md [hunk 6]
```diff
@@ -412,7 +412,7 @@ es). It should look like this:
 
 ```yaml
 form-flow:
-  inputs: 'org.formflowstartertemplate.app.inputs.'
+  inputs: 'formflow.library.inputs.'
```
Comment (bseeger) [inline README.md:415]: This doesn't seem right. It should probably be more like `your.apps.path.inputs` or something like that, right?  This needs to be specific to an implementors path, I think.

File: src/main/java/formflow/library/config/ActionManager.java [hunk 7]
```diff
@@ -166,14 +166,15 @@ public void handleBeforeDisplayAction(ScreenNavigationConfiguration currentScree
    *
    * @param currentScreen  The screen that we are currently saving data from.
    * @param formSubmission The current form submission
+   * @param submission     The submission object after changes to the current screen have been saved to the repository
```
Comment (bseeger) [inline src/main/java/formflow/library/config/ActionManager.java:169]: Validation is run before we merge the data into the Submission.  So the Submission does not have the currently submitted FormSubmission data yet.   The current Submission here is what is currently in the DB _before_ the current screens data has been merged into it.   That's my understanding at least, is yours different?

File: README.md (hunk 6 above)
Comment (bseeger) [reply README.md:415]: This is a tricky spot to show an example... not quite sure how to do it.

File: src/main/java/formflow/library/config/ActionManager.java (hunk 7 above)
Comment (coltborg) [reply src/main/java/formflow/library/config/ActionManager.java:169]: Yes, you are totally right! Will update.

File: README.md (hunk 6 above)
Comment (coltborg) [reply README.md:416]: How about:

```suggestion
  inputs: '<your-apps-custom-path-to-inputs>.inputs.'
```

Comment (bseeger) [reply README.md]: I think that would work, it makes it clear that you need to do something there.  I'm not sure there's a good way to do this, so I'd be fine with what you've crafted.

**Review by bseeger — APPROVED** [review]
Looks good to me!

**Pull request merged**